File: sdk/platform/Web/src/PoseEstimation.cpp
#include <Posemesh/PoseEstimation.hpp>
#include <emscripten/bind.h>
#include <iostream>
#include <memory>
#include <vector>

using namespace emscripten;
using namespace psm;

namespace {
Pose solvePnP(const std::vector<std::shared_ptr<Landmark>>& landmarks,
    const std::vector<std::shared_ptr<LandmarkObservation>>& landmarkObservations,
    const Matrix3x3& cameraMatrix,
    int method)
{
    std::vector<Landmark> landmarksRaw;
    for (int i = 0; i < landmarks.size(); ++i) {
        landmarksRaw.push_back(*landmarks[i]);
    }

    std::vector<LandmarkObservation> landmarkObservationsRaw;
    for (int i = 0; i < landmarkObservations.size(); ++i) {
        landmarkObservationsRaw.push_back(*landmarkObservations[i]);
    }

    return PoseEstimation::solvePnP(landmarksRaw, landmarkObservationsRaw, cameraMatrix, (psm::SolvePnpMethod)method);
}
}

EMSCRIPTEN_BINDINGS(PoseEstimation)
{
    class_<PoseEstimation>("PoseEstimation")
        .class_function("__solvePnP(landmarks, landmarkObservations, cameraMatrix, method)", &solvePnP);
}

File: sdk/platform/Web/PoseEstimation.js
posemeshModule.PoseEstimation = null;

__internalPosemeshAPI.builderFunctions.push(function() {
    __internalPosemesh.PoseEstimation.solvePnP = function(landmarks, landmarkObservations, cameraMatrix, method) {
        let landmarksVector = undefined, landmarkObservationsVector = undefined;
        try {
            landmarksVector = __internalPosemeshAPI.toVectorLandmark(landmarks, false);
            landmarkObservationsVector = __internalPosemeshAPI.toVectorLandmarkObservation(landmarkObservations, false);
            return __internalPosemesh.PoseEstimation.__solvePnP(landmarksVector, landmarkObservationsVector, cameraMatrix, method);
        } finally {
            if (landmarkObservationsVector) {
                landmarkObservationsVector.delete();
            }
            if (landmarksVector) {
                landmarksVector.delete();
            }
        }
    };

    posemeshModule.PoseEstimation = __internalPosemesh.PoseEstimation;
});
